File: data/spectrogram.c
#include "spectrogram.h"
#include <stdio.h>
#include <math.h>
#include <string.h>

// Function to save debug output to a file
void save_debug_output(const char* filename, const char* message, float* data, int rows, int cols) {
    FILE* file = fopen(filename, "a");
    if (!file) {
        printf("Error: Could not open debug file for writing\n");
        return;
    }
    fprintf(file, "%s\n", message);
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            fprintf(file, "%.6f ", data[i * cols + j]);
        }
        fprintf(file, "\n");
    }
    fprintf(file, "\n");
    fclose(file);
}

void pre_emphasis(short* audio, float pre_emphasis_array[], unsigned long num_samples) {
    float norm[num_samples];
    for (int i = 0; i < num_samples; i++) {
        norm[i] = audio[i] / 32768.0f;
    }
    pre_emphasis_array[0] = norm[0];
    for (int i = 1; i < num_samples; i++) {
        pre_emphasis_array[i] = norm[i] - COEFFICIENT * norm[i - 1];
    }
    // Save debug output
    // save_debug_output("debug.txt", "Pre-emphasis output:", pre_emphasis_array, 1, num_samples);
}

void apply_windowing(float* frame, int size) {
    for (int i = 0; i < size; i++) {
        float window = 0.54f - 0.46f * cos(2 * M_PI * i / (size - 1));
        frame[i] *= window;
    }
    // Save debug output
    float windowed_frame[size];
    for (int i = 0; i < size; i++) {
        windowed_frame[i] = frame[i];  // Save real part for debugging
    }
    //save_debug_output("debug.txt", "Windowed frame output:", windowed_frame, 1, size);
}

void spectrogram_population(kiss_fft_cpx fft_out[], float spectrogram[], int frame) {
    for (int i = 0; i < NUM_BINS; i++) {
        spectrogram[frame * NUM_BINS + i] = sqrt(fft_out[i].r * fft_out[i].r + fft_out[i].i * fft_out[i].i);
    }
    // Save debug output
    //save_debug_output("debug.txt", "Spectrogram population output:", spectrogram + frame * NUM_BINS, 1, NUM_BINS);
}

void framing_operation(float* pre_emphasis_audio, float spectrogram[], unsigned long num_samples) {
    kiss_fftr_cfg cfg = kiss_fftr_alloc(FRAME_SIZE, 0, NULL, NULL);
    kiss_fft_cpx fft_out[FRAME_SIZE/2+1];
    float fft_in[FRAME_SIZE];

    for (int frame = 0; frame < NUM_FRAMES(num_samples); frame++) {
        int start = frame * FRAME_STRIDE;
        for (int i = 0; i < FRAME_SIZE; i++) {
            fft_in[i] = (i < FRAME_SIZE) ? pre_emphasis_audio[start + i] : 0.0f;
        }
        apply_windowing(fft_in, FRAME_SIZE);
        kiss_fftr(cfg, fft_in, fft_out);
        spectrogram_population(fft_out, spectrogram, frame);
    }
    free(cfg);
    // Save debug output
    //save_debug_output("debug.txt", "Framing operation output:", spectrogram, NUM_FRAMES(num_samples), NUM_BINS);
}

float hz_to_mel(float hz) {
    return 1127.0f * log10(1+hz/700+1e-20);
}

float mel_to_hz(float mel) {
    return 700 * (pow(10,mel/1127.0f) - 1);
}

void triangle(float* out, float* x, int size, int left, int middle, int right) {
    for(int i=0; i<size; i++) {
        if(x[i]<=left || x[i]>=right) {
            out[i]=0.0f;
        }
        else if (x[i]<=middle) {
            out[i]=(x[i]-left)/(middle-left);
        }
        else {
            out[i]=(right-x[i])/(right-middle);
        }
        if (!isfinite(out[i])) out[i] = 0.0f;
        //printf("%.6f\t", out[i]);
    }
}

void create_mel_filterbank(float mel_filterbank[FILTER_NUMBER][NUM_BINS]) {
    float min_mel = hz_to_mel(MIN_FREQ); //0
    float max_mel = hz_to_mel(MAX_FREQ); //1233,4127
    float mel_spacing=(max_mel-min_mel)/(FILTER_NUMBER+1);
    float mel_points[FILTER_NUMBER + 2];
    for (int i = 0; i < FILTER_NUMBER + 2; i++) {
        float value=min_mel+i*mel_spacing;
        if(value>max_mel) {
            value=max_mel;
        }
        mel_points[i]= mel_to_hz(value);
        if(mel_points[i]>MAX_FREQ) {
            mel_points[i]=MAX_FREQ;
        }
    }
    for (int i=0; i < FILTER_NUMBER; i++) {
        for(int j=0; j<NUM_BINS; j++) {
            mel_filterbank[i][j]=0.0f;
        }
        float left=mel_points[i];
        float middle=mel_points[i+1];
        float right=mel_points[i+2];

        int left_bin = (int)(left * (NUM_BINS - 1) / (SAMPLE_RATE / 2));
        int middle_bin = (int)(middle * (NUM_BINS - 1) / (SAMPLE_RATE / 2));
        int right_bin = (int)(right * (NUM_BINS - 1) / (SAMPLE_RATE / 2));

        left_bin = max(0, min(NUM_BINS-1, left_bin));
        middle_bin = max(0, min(NUM_BINS-1, middle_bin));
        right_bin = max(0, min(NUM_BINS-1, right_bin));

        if (left_bin==middle_bin) middle_bin = min(left_bin+1, NUM_BINS-1);
        if (middle_bin==right_bin) right_bin = min(middle_bin+1, NUM_BINS-1);

        int size = right_bin - left_bin + 1;
        float z[size];
        for(int j=0; j<size; j++) {
            z[j]=left_bin+j;
        }
        float out[size];
        triangle(out, z, size, left_bin, middle_bin, right_bin);
        for(int j=0; j<size; j++) {
            mel_filterbank[i][left_bin+j]=out[j];
        }
    }
    // Save debug output
    //save_debug_output("debug.txt", "Mel filterbank output:", (float*)mel_filterbank, FILTER_NUMBER, NUM_BINS);
}

void apply_mel_filterbank(float spectrogram[], float mel_filterbank[FILTER_NUMBER][NUM_BINS], float log_mel_spectrogram[], int num_frames) {
    for (int i = 0; i < num_frames; i++) {
        for (int j = 0; j < FILTER_NUMBER; j++) {
            float sum = 0.0f;
            for (int k = 0; k < NUM_BINS; k++) {
                sum += spectrogram[i * NUM_BINS + k] * mel_filterbank[j][k];
            }
            log_mel_spectrogram[i * FILTER_NUMBER + j] = 10*log10(sum+1e-20);
        }
    }
    // Save debug output
    //save_debug_output("debug.txt", "Log Mel spectrogram output:", log_mel_spectrogram, num_frames, FILTER_NUMBER);
}

void save_spectrogram(float log_mel_spectrogram[], int num_frames, const char* filename) {
    FILE* file = fopen(filename, "w");
    if (!file) {
        printf("Error: Could not open file for writing\n");
        return;
    }
    for (int frame = 0; frame < num_frames; frame++) {
        for (int i = 0; i < FILTER_NUMBER; i++) {
            fprintf(file, "%.6f ", log_mel_spectrogram[frame * FILTER_NUMBER + i]);
        }
        fprintf(file, "\n");
    }
   /*for(int i=0; i<num_frames; i++) {
        for(int j=0; j<FILTER_NUMBER; j++) {
            fprintf(file, "%.6f\t%.6f\n", log_mel_spectrogram[i * FILTER_NUMBER + j], spectrogram_sample[i * FILTER_NUMBER + j]);
        }
    }*/
    fclose(file);
}

void mean_filter(float log_mel_spectrogram[], int num_frames) {
    for (int i = 0; i < num_frames; i++) {
        for (int j = 0; j < FILTER_NUMBER; j++) {
            float sum = 0.0f;
            int count = 0;
            for (int di = -1; di <= 1; di++) {
                for (int dj = -1; dj <= 1; dj++) {
                    int ni = i + di;
                    int nj = j + dj;
                    if (ni >= 0 && ni < num_frames && nj >= 0 && nj < FILTER_NUMBER) {
                        sum += log_mel_spectrogram[ni * FILTER_NUMBER + nj];
                        count++;
                    }
                }
            }
            log_mel_spectrogram[i * FILTER_NUMBER + j] = sum / count;
        }
    }
    // Save debug output
    //save_debug_output("debug.txt", "Mean filter output:", log_mel_spectrogram, num_frames, FILTER_NUMBER);
}

void apply_noise_floor(float log_mel_spectrogram[], int num_frames) {
    for (int i = 0; i < num_frames; i++) {
        for (int j = 0; j < FILTER_NUMBER; j++) {
            float value = log_mel_spectrogram[i * FILTER_NUMBER + j];
            value = (value - NOISE_FLOOR) / ((-1 * NOISE_FLOOR) + 12);
            if (value < 0) value = 0;
            if (value > 1) value = 1;
            int quantized = (int)round(value * 256);
            if (quantized < 0) quantized = 0;
            if (quantized > 255) quantized = 255;
            log_mel_spectrogram[i * FILTER_NUMBER + j] = (float)quantized / 256.0f;
            log_mel_spectrogram[i*FILTER_NUMBER+j]=(log_mel_spectrogram[i*FILTER_NUMBER+j]>=0.65) ? log_mel_spectrogram[i*FILTER_NUMBER+j] : 0.0f;
        }
    }
    // Save debug output
    //save_debug_output("debug.txt", "Noise floor output:", log_mel_spectrogram, num_frames, FILTER_NUMBER);
}

void compute_spectrogram(short* audio, float log_mel_spectrogram[], unsigned long num_samples) {
    float pre_emphasis_array[num_samples];
    pre_emphasis(audio, pre_emphasis_array, num_samples);

    float spectrogram[NUM_FRAMES(num_samples) * NUM_BINS];
    framing_operation(pre_emphasis_array, spectrogram, num_samples);

    float mel_filterbank[FILTER_NUMBER][NUM_BINS];
    create_mel_filterbank(mel_filterbank);

    apply_mel_filterbank(spectrogram, mel_filterbank, log_mel_spectrogram, NUM_FRAMES(num_samples));

    // Uncomment to apply mean filter
    //mean_filter(log_mel_spectrogram, NUM_FRAMES(num_samples));

    apply_noise_floor(log_mel_spectrogram, NUM_FRAMES(num_samples));

    save_spectrogram(log_mel_spectrogram, NUM_FRAMES(num_samples), SPECTROGRAM_FILENAME);
}

File: data/spectrogram.py
import numpy as np
import matplotlib.pyplot as plt
import librosa
from scipy.fftpack import dct
from scipy.signal import get_window

SAMPLE_RATE = 16000
FRAME_DUR = 0.032
FRAME_SIZE = int(SAMPLE_RATE * FRAME_DUR)
FRAME_STRIDE_DUR = 0.024
FRAME_STRIDE = int(SAMPLE_RATE * FRAME_STRIDE_DUR)
NUM_BINS = FRAME_SIZE // 2
FILTER_NUMBER = 40
MIN_FREQ = 0
MAX_FREQ = SAMPLE_RATE // 2
COEFFICIENT = 0.96875
NOISE_FLOOR = -40.0

def pre_emphasis(audio):
    emphasized = np.zeros_like(audio, dtype=np.float32)
    emphasized[0] = audio[0] / 32768.0
    for i in range(1, len(audio)):
        emphasized[i] = (audio[i] / 32768.0) - COEFFICIENT * (audio[i-1] / 32768.0)
    return emphasized

def apply_windowing(frame):
    window = 0.54 - 0.46 * np.cos(2 * np.pi * np.arange(len(frame)) / (len(frame) - 1))
    return frame * window

def hz_to_mel(hz):
    return 1127.0 * np.log10(1 + hz / 700.0)

def mel_to_hz(mel):
    return 700 * (10 ** (mel / 1127.0) - 1)

def create_mel_filterbank():
    min_mel = hz_to_mel(MIN_FREQ)
    max_mel = hz_to_mel(MAX_FREQ)
    #mel_points = np.linspace(min_mel, max_mel, FILTER_NUMBER + 2)
    #hz_points = mel_to_hz(mel_points)
    mel_points = np.zeros(FILTER_NUMBER + 2)
    mel_spacing = (max_mel - min_mel) / (FILTER_NUMBER + 1)
    for i in range(FILTER_NUMBER + 2):
        mel_points[i] = mel_to_hz(min_mel + i * mel_spacing)
        if mel_points[i] > MAX_FREQ:
            mel_points[i] = MAX_FREQ

    #bin_indices = np.floor((NUM_BINS) * hz_points / (SAMPLE_RATE / 2)).astype(int)
    #bin_indices = np.clip(bin_indices, 0, NUM_BINS - 1)
    bin_indices = np.zeros(FILTER_NUMBER + 2, dtype=int)
    for i in range(FILTER_NUMBER + 2):
        bin_indices[i] = int(mel_points[i] * (NUM_BINS - 1) / (SAMPLE_RATE / 2.0))
        bin_indices[i] = max(0, min(NUM_BINS - 1, bin_indices[i]))

    filterbank = np.zeros((FILTER_NUMBER, NUM_BINS))

    for i in range(FILTER_NUMBER):
        left = bin_indices[i]
        middle = bin_indices[i+1]
        right = bin_indices[i+2]

        if left == middle:
            middle = min(left + 1, NUM_BINS - 1)
        if middle == right:
            right = min(middle + 1, NUM_BINS - 1)

        #filterbank[i, left:middle] = np.linspace(0, 1, middle - left)
        for j in range(left, middle):
            filterbank[i, j] = (j - left) / (middle - left)

        #filterbank[i, middle:right] = np.linspace(1, 0, right - middle)
        for j in range(middle, right):
            filterbank[i, j] = 1.0 - (j - middle) / (right - middle)
    return filterbank

def compute_spectrogram(audio, show_plot=True):
    num_samples = len(audio)

    total_duration = num_samples / SAMPLE_RATE
    num_frames_full_second = int((total_duration - FRAME_DUR) / FRAME_STRIDE_DUR) + 1
    num_frames = min(num_frames_full_second, 40)
    pre_emphasis_array = pre_emphasis(audio)
    spectrogram = np.zeros((num_frames, NUM_BINS))

    for frame in range(num_frames):
        start = frame * FRAME_STRIDE
        end = start + FRAME_SIZE
        segment = pre_emphasis_array[start:end]
        if len(segment) < FRAME_SIZE:
            segment = np.pad(segment, (0, FRAME_SIZE - len(segment)))

        windowed = apply_windowing(segment)
        fft = np.fft.rfft(windowed, n=FRAME_SIZE)
        magnitude = np.abs(fft)
        spectrogram[frame] = magnitude[:NUM_BINS]

    mel_filterbank = create_mel_filterbank()
    mel_spectrogram = np.dot(spectrogram, mel_filterbank.T)
    log_mel_spectrogram = 10 * np.log10(mel_spectrogram + 1e-20)

    log_mel_spectrogram = (log_mel_spectrogram - NOISE_FLOOR) / (-NOISE_FLOOR + 12)
    log_mel_spectrogram = np.clip(log_mel_spectrogram, 0, 1)
    quantized = np.round(log_mel_spectrogram * 256) / 256.0
    quantized = np.where(quantized >= 0.65, quantized, 0)
    quantized = quantized[:40]

    if show_plot:
        plt.figure(figsize=(10, 6))
        time_axis = np.linspace(0, 0.968, 40)
        plt.imshow(quantized.T, aspect='auto', origin='lower',
                  extent=[0, 0.968, 0, FILTER_NUMBER])
        plt.colorbar(label='Magnitude')
        plt.xlabel('Time (s)')
        plt.ylabel('Mel filter index')
        plt.title('40x40 Mel Spectrogram (0.968s duration)')
        plt.show()

    return quantized

if __name__ == "__main__":
    audio_file = <Insert WAV FILE>
    audio, sr = librosa.load(audio_file, sr=SAMPLE_RATE, mono=True)

    audio_int16 = (audio * 32768).astype(np.int16)

    required_samples = int(0.968 * SAMPLE_RATE)
    audio_int16 = audio_int16[:required_samples]

    mel_spectrogram = compute_spectrogram(audio_int16)
    print(f"Spectrogram shape: {mel_spectrogram.shape}")

    with open('mfe_features_formatted.txt', 'w') as f:
      for row in mel_spectrogram:
          # Write each row with 6 decimal places, space-separated
          line = ' '.join([f'{x:.6f}' for x in row])
          f.write(line + '\n')
